pennylane_lightning/src/util/Memory.hpp

```diff
@@ -0,0 +1,197 @@
+// Copyright 2022 Xanadu Quantum Technologies Inc.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#pragma once
+
+#include <algorithm>
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include <cstdlib>
+#include <limits>
+#include <memory>
+#include <new>
+
+#include "BitUtil.hpp"
+#include "TypeList.hpp"
+
+namespace Pennylane::Util {
+/**
+ * @brief Custom aligned allocate function.
+ *
+ * We use `posix_memalign` for MacOS as Mac does not support
+ * `std::aligned_alloc` properly yet (even in MacOS 10.15).
+ * Note that alignment must be larger than max_align_t. Otherwise, the behavior
+ * is undefined.
+ *
+ * @param alignment Alignment value we want for the data pointer
+ * @param bytes Number of bytes to allocate
+ * @return Pointer to the allocated memory
+ */
+inline auto alignedAlloc(uint32_t alignment, size_t bytes) -> void * {
+#if defined(__clang__) && defined(__APPLE__)
+    /* Apple clang does not support std::aligned_alloc in Mac 10.14.
```

Comment:
Since we removed MacOS 10.14 support, should we be able to remove this?

Reply:
Huh, that's really odd. Sure, keeping this is fine.
If the level of support ever becomes a problem we can build MacOS wheels from GCC or clang mainline compilers rather than Apple.

Reply:
Yes, I would prefer GCC as it supports new C++ functionalities better and will reduce the maintenance cost (by using the same compiler for Linux and MacOS). I may try this after this PR.